test/Microsoft.ML.Tests/Transformers/HashTests.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.RunTests;
+using Microsoft.ML.Runtime.Tools;
+using Microsoft.ML.Transforms;
+using System.IO;
+using System.Linq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests.Transformers
+{
+    public class HashTests : TestDataPipeBase
+    {
+        public HashTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        private class TestClass
+        {
+            public float A;
+            public float B;
+            public float C;
+        }
+
+        private class TestMeta
+        {
+            [VectorType(2)]
+            public float[] A;
+            public float B;
+            [VectorType(2)]
+            public double[] C;
+            public double D;
+        }
+
+        [Fact]
+        public void HashWorkout()
+        {
+            var data = new[] { new TestClass() { A = 1, B = 2, C = 3, }, new TestClass() { A = 4, B = 5, C = 6 } };
+
+            var dataView = ComponentCreation.CreateDataView(Env, data);
+            var pipe = new HashConverter(Env, new[]{
+                    new HashConverterTransformer.ColumnInfo("A", "HashA", hashBits:4, invertHash:-1),
+                    new HashConverterTransformer.ColumnInfo("B", "HashB", hashBits:3, ordered:true),
+                    new HashConverterTransformer.ColumnInfo("C", "HashC", seed:42),
+                    new HashConverterTransformer.ColumnInfo("A", "HashD"),
+                });
+
+            TestEstimatorCore(pipe, dataView);
+            Done();
+        }
+
+        [Fact]
+        public void TestMetadata()
+        {
+
+            var data = new[] {
+                new TestMeta() { A=new float[2] { 3.5f, 2.5f}, B=1, C= new double[2] { 5.1f, 6.1f}, D= 7},
+                new TestMeta() { A=new float[2] { 3.5f, 2.5f}, B=1, C= new double[2] { 5.1f, 6.1f}, D= 7},
+                new TestMeta() { A=new float[2] { 3.5f, 2.5f}, B=1, C= new double[2] { 5.1f, 6.1f}, D= 7}};
+
+
+            var dataView = ComponentCreation.CreateDataView(Env, data);
+            var pipe = new HashConverter(Env, new[] {
+                new HashConverterTransformer.ColumnInfo("A", "HashA", invertHash:1, hashBits:10),
```

Comment:
>HashConverterTransformer [](start = 20, length = 24)

hmm, now that I think of it, we should probably have `ColumnInfo` be part of the estimator as well, right?

Reply:
Do you want this to be part of this PR or as separate PR?

---
In reply to: [218936515](https://github.com/dotnet/machinelearning/pull/944#discussion_r218936515) [](ancestors = 218936515)

Reply:
this PR, if not too cumbersome

---
In reply to: [218941758](https://github.com/dotnet/machinelearning/pull/944#discussion_r218941758) [](ancestors = 218941758,218936515)

Reply:
I mean, we have this pattern in all estimators. Wouldn't it be better to move stuff in one PR, rather than in separate? Or you want me to update all estimators as well?

---
In reply to: [218944749](https://github.com/dotnet/machinelearning/pull/944#discussion_r218944749) [](ancestors = 218944749,218941758,218936515)

Reply:
I would rather prefer to keep one style in code, and if it needed to flip everything, other than have two styles in code.

---
In reply to: [218945227](https://github.com/dotnet/machinelearning/pull/944#discussion_r218945227) [](ancestors = 218945227,218944749,218941758,218936515)

Reply:
Fair. Let's do it in a different PR for all at once, where applicable

---
In reply to: [218954028](https://github.com/dotnet/machinelearning/pull/944#discussion_r218954028) [](ancestors = 218954028,218945227,218944749,218941758,218936515)